src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -357,6 +358,159 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncode(char* s, int length)
+        {
+            int numUnsafe = 0;
+            int numSpaces = 0;
+            for (int i = 0; i < length; i++)
+            {
+                char c = s[i];
+                if (c == ' ')
+                {
+                    numSpaces++;
+                }
+                else if (!IsUrlSafeChar(s[i]))
+                {
+                    numUnsafe += GetByteCount(s, i, length);
+                }
+            }
+
+            if (numSpaces == 0 && numUnsafe == 0)
+            {
+                // Nothing to expand
+                return new string(s, 0, length);
+            }
+
+            byte[] originalBytes = GetBytes(s, length);
```

Comment:
Why did you implement GetBytes manually rather than allocating an array of the right length and using Encoding.UTF8.GetBytes that takes a char*?